deepchem/models/losses.py

```diff
@@ -465,6 +468,119 @@ def loss(inputs):
+            av_task_dist_loss = nn.MSELoss(reduction="mean")
+            fg_task_dist_loss = nn.MSELoss(reduction="mean")
+
+            sigmoid = nn.Sigmoid()
+
+            av_atom_loss = av_task_loss(preds['av_task'][0], targets["av_task"])
+            av_bond_loss = av_task_loss(preds['av_task'][1], targets["av_task"])
+            bv_atom_loss = av_task_loss(preds['bv_task'][0], targets["bv_task"])
+            bv_bond_loss = av_task_loss(preds['bv_task'][1], targets["bv_task"])
+
+            fg_atom_from_atom_loss = fg_task_loss(
+                preds["fg_task"]["atom_from_atom"], targets["fg_task"])
+            fg_atom_from_bond_loss = fg_task_loss(
+                preds["fg_task"]["atom_from_bond"], targets["fg_task"])
+            fg_bond_from_atom_loss = fg_task_loss(
+                preds["fg_task"]["bond_from_atom"], targets["fg_task"])
+            fg_bond_from_bond_loss = fg_task_loss(
+                preds["fg_task"]["bond_from_bond"], targets["fg_task"])
+
+            av_dist_loss = av_task_dist_loss(preds['av_task'][0],
+                                             preds['av_task'][1])
+            bv_dist_loss = av_task_dist_loss(preds['bv_task'][0],
+                                             preds['bv_task'][1])
+
+            fg_atom_dist_loss = fg_task_dist_loss(
+                sigmoid(preds["fg_task"]["atom_from_atom"]),
+                sigmoid(preds["fg_task"]["atom_from_bond"]))
+            fg_bond_dist_loss = fg_task_dist_loss(
+                sigmoid(preds["fg_task"]["bond_from_atom"]),
+                sigmoid(preds["fg_task"]["bond_from_bond"]))
+
+            av_bv_loss = av_atom_loss + av_bond_loss + bv_atom_loss + bv_bond_loss
+            fg_loss = fg_atom_from_atom_loss + fg_atom_from_bond_loss + fg_bond_from_atom_loss + fg_bond_from_bond_loss
+            fg_dist_loss = fg_atom_dist_loss + fg_bond_dist_loss
+
+            # NOTE The below comment is from original source code
+            # dist_loss = av_dist_loss + bv_dist_loss + fg_dist_loss
+            # return av_loss + fg_loss + dist_coff * dist_loss
+            overall_loss = av_bv_loss + fg_loss + dist_coff * (
+                av_dist_loss + bv_dist_loss + fg_dist_loss)
+
+            # return overall_loss, av_loss, bv_loss, fg_loss, av_dist_loss, bv_dist_loss, fg_dist_loss
+            # We just return overall_loss since TorchModel can handle only a single loss
+            return overall_loss
+
+        return loss
+
+
+class GroverFinetuneLoss(Loss):
+
+    def _create_pytorch_loss(self):
+        # TODO It should be BCEWithLogitsLoss if dataset type is classification.
+        # We are assuming here task type to be regression.
+        import torch.nn as nn
+
+        def loss(preds, targets, dist_coff=0.1):
+            pred_loss = nn.MSELoss()
+
+            # print(type(preds))
+            # TODO: Here, should we need to involve the model status? Using len(preds) is just a hack.
+            if type(preds) is not tuple:
```

Comment:
What is the expected structure of `preds` here? We shouldn't make use of undocumented structure otherwise will be hard to maintain

Reply:
The finetune model in training mode produces a set of output and in evaluation model it produces a different set of outputs. So, it can be a tuple or just the predictions.